### PR TITLE
Refactor and introduce new configuration options

### PR DESCRIPTION
**Pull request by zenoone**
This pull request refactors the wave function ansatz and implements new configuration options to extend the wave fucntion design space. Furthermore, it reduces ambiguities by removing ansatz defaults from the code and extending the hydra configs. With that provides config files for an ansatz close to PauliNet and FermiNet respectively.

### CODE REVIEW

File: src/deepqmc/wf/paulinet/paulinet.py [hunk 1]
```diff
@@ -46,79 +44,51 @@ def eval_log_slater(xs):
...
-        confs (int, (:math:`N_\text{conf}`, :math:`N_\text{elec}`)): optional,
-            specifies the electronic configuration, i.e. the occupied orbitals
-            in each determinant. The value of :data:`confs[i,j]` is the index of
-            the orbital occupied by the :data:`j` th electron in the :data:`i` th
-            determinant.
-        backflow_op (Callable): optional, specifies how the backflow is applied
-            to the orbitals.
-        full_determinant (bool): optional, if :data:`False`, the determinants are
+        hamil (~MolecularHamiltonian): the Hamiltonian of the system.
+        basis (~wf.paulinet.env.ExponentialEnvelopes): the orbital envelopes.
```
Comment (szbernat) [inline src/deepqmc/wf/paulinet/paulinet.py:51]: Maybe its time to rename this, e.g. to `envelope`, since this has not much to do anymore with an AO basis.

**Review by szbernat — COMMENTED** [review]
It's nice to see all the removed complexity!

File: src/deepqmc/gnn/schnet.py [hunk 2]
```diff
@@ -3,164 +3,178 @@
 from jax import ops
 
 from ..hkext import MLP
-from .edge_features import PauliNetEdgeFeatures
+from ..utils import flatten
 from .gnn import GraphNeuralNetwork, MessagePassingLayer
 from .graph import GraphNodes, difference_callback
 
 
 class SchNetLayer(MessagePassingLayer):
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/gnn/schnet.py:11]: Yes, I went over the PR again and the diff seemed more transparent if we do this later. We can finish the review and then add a commit that changes SchNet to ElectronGNN and PauliNet to NeuralNetworkWaveFunction.

File: src/deepqmc/gnn/schnet.py [hunk 3]
```diff
@@ -172,24 +186,37 @@ def aggregate_edges_for_nodes(nodes, edges):
...
+                'same': z['same'],
+                'anti': z['anti'],
+                'ee': z['same'] + z['anti'],
+                'ne': z['ne'],
+            }
+            f = {typ: FEATURE_MAPPING[typ] for typ in self.update_features}
+            if self.update_rule == 'concatenate':
+                updated = self.g(jnp.concatenate(jnp.stack(list(f.values())), axis=-1))
+            elif self.update_rule == 'featurewise':
+                updated = jnp.sum(
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/gnn/schnet.py:210]: Yes, great suggestion. We just have to make sure that the option where we sum first retains the dimension.

File: src/deepqmc/wf/paulinet/env.py [hunk 4]
```diff
@@ -10,11 +12,20 @@ def __init__(self, shells):
...
     def __call__(self, diffs):
-        return jnp.exp(-jnp.abs(self.zetas * jnp.sqrt(diffs[..., self.center_idx, -1])))
+        return jnp.exp(
+            -norm(
+                jnp.einsum(
+                    'ers,ies->ier', self.zetas, diffs[..., self.center_idx, :-1]
+                ),
+                safe=True,
+            )
+        )
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/wf/paulinet/env.py:28]: Yes, this was on my todolist. Lets have both the isotropic and non-isotropic envelopes.

File: src/deepqmc/wf/paulinet/paulinet.py (hunk 1 above)
Comment (zenoone) [reply src/deepqmc/wf/paulinet/paulinet.py:51]: Yes I agree. Basis might be confusing.

File: src/deepqmc/conf/ansatz/ferminet.yaml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        _target_: deepqmc.hkext.ssp
+        _partial_: true
+ 
+  gnn_factory:
+    _target_: deepqmc.gnn.schnet.SchNet
+    _partial_: true
+    n_interactions: 4
+    positional_electron_embeddings: true
+    atom_type_embeddings: false
+    edge_feat_factory:
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/conf/ansatz/ferminet.yaml:35]: I agree. The EdgeFeatures are a bit convoluted for what we want to achieve with them and another iteration would suit them well.

File: src/deepqmc/conf/ansatz/ferminet.yaml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      _target_: deepqmc.gnn.edge_features.EdgeFeatures
+      _partial_: true
+      n_rbf: 1
+      r_cut: 30.
+      irreps: 0y+1y
+      equivariant: false
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/conf/ansatz/ferminet.yaml:41]: Yes, I think the equivariant option should be removed for now. I would argue that for what we usually use for the edge features there is no need to have the `e3nn-jax` depndency and the extrafunctionality could be removed or refactored not to depend on it.

File: src/deepqmc/conf/ansatz/ferminet.yaml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  embedding_dim: 256
+  jastrow_factory: null
+  backflow_factory:  
+    _target_: deepqmc.wf.paulinet.omni.Backflow
+    _partial_: true
+    subnet_kwargs:
+      hidden_layers: ['log', 1]
+      last_linear: true
+      activation: 
+        _target_: deepqmc.hkext.ssp
```
[the remark this answers is not in the document]
Comment (zenoone) [reply src/deepqmc/conf/ansatz/ferminet.yaml:26]: As disucssed the `OmegaConf` resolver has to be called somewhere in the code. However, we wanted to refrain from executing code on the deepqmc import. With this solution we aviod calling the custom resolver on import and the increased verbosity did not seem to harmful, given that this is a reoccuring pattern from the module initialization.

File: src/deepqmc/conf/ansatz/ferminet.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (szbernat) [reply src/deepqmc/conf/ansatz/ferminet.yaml]: I'll just remove the `PolynomialEnvelope` and `BesselBasis` for now, thus get rid of the `e3nn-jax` dependency.

File: src/deepqmc/gnn/schnet.py [hunk 8]
```diff
@@ -172,24 +186,37 @@ def aggregate_edges_for_nodes(nodes, edges):
...
+                    .repeat(self.n_up + self.n_down, axis=0)
+                ),
+                'nodes_down': (
+                    nodes.electrons[self.n_up :]
+                    .mean(axis=0, keepdims=True)
+                    .repeat(self.n_up + self.n_down, axis=0)
+                ),
+                'same': z['same'],
+                'anti': z['anti'],
+                'ee': z['same'] + z['anti'],
```
[the remark this answers is not in the document]
Comment (szbernat) [reply src/deepqmc/gnn/schnet.py:203]: Feel free to resolve this conversion, just maybe keep this in mind

File: src/deepqmc/gnn/schnet.py [hunk 9]
```diff
@@ -172,24 +186,37 @@ def aggregate_edges_for_nodes(nodes, edges):
...
-                else:
-                    zs = jnp.concatenate(
-                        [nodes.electrons, z['same'] + z['anti'], z['ne']],
-                        axis=-1,
-                    )
-                    updated = self.g(zs)
-            else:
-                gs = jnp.stack(
-                    [self.g[typ](z[typ]) for typ in self.edge_types], axis=-1
+            FEATURE_MAPPING = {
```
[the remark this answers is not in the document]
Comment (szbernat) [reply src/deepqmc/gnn/schnet.py:189]: Feel free to close this conversation, we should just keep this in mind.

**Comment by zenoone** [comment]
The PR currently breaks the pseudo potentials. This should be fixed before merging.

**Pull request merged**